library/cipher.c

```diff
@@ -404,9 +290,375 @@ int mbedtls_cipher_setkey(mbedtls_cipher_context_t *ctx,
         return mbedtls_cipher_get_base(ctx->cipher_info)->setkey_dec_func(ctx->cipher_ctx, key,
                                                                           ctx->key_bitlen);
     }
-
-    return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+
+    return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+}
+
+int mbedtls_cipher_update(mbedtls_cipher_context_t *ctx, const unsigned char *input,
+                          size_t ilen, unsigned char *output, size_t *olen)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t block_size;
+
+    if (ctx->cipher_info == NULL) {
+        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+    }
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO) && !defined(MBEDTLS_DEPRECATED_REMOVED)
+    if (ctx->psa_enabled == 1) {
+        /* While PSA Crypto has an API for multipart
+         * operations, we currently don't make it
+         * accessible through the cipher layer. */
+        return MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE;
+    }
+#endif /* MBEDTLS_USE_PSA_CRYPTO && !MBEDTLS_DEPRECATED_REMOVED */
+
+    *olen = 0;
+    block_size = mbedtls_cipher_get_block_size(ctx);
+    if (0 == block_size) {
+        return MBEDTLS_ERR_CIPHER_INVALID_CONTEXT;
+    }
+
+    if (((mbedtls_cipher_mode_t) ctx->cipher_info->mode) == MBEDTLS_MODE_ECB) {
+        if (ilen != block_size) {
+            return MBEDTLS_ERR_CIPHER_FULL_BLOCK_EXPECTED;
+        }
+
+        *olen = ilen;
+
+        if (0 != (ret = mbedtls_cipher_get_base(ctx->cipher_info)->ecb_func(ctx->cipher_ctx,
+                                                                            ctx->operation, input,
+                                                                            output))) {
+            return ret;
+        }
+
+        return 0;
+    }
+
+#if defined(MBEDTLS_GCM_C)
```

Comment:
I think this should be guarded with `&& defined(CIPHER_C)` as well.

Reply:
I actually skipped that on purpose because I saw `mbedtls_cipher_update` was used by both CCM and GCM so I thought this was necessary for them. Functions like `mbedtls_ccm_crypt` and `mbedtls_ccm_update_ad` use `mbedtls_cipher_update` which eventually returns back to `mbedtls_ccm_update`. I suspect that removing that part might prevent these functions to work properly. Or (option B) we must prevent using cipher functions, so instead of `ccm->cipher->ccm`, we just do `ccm->ccm`.

Reply:
And when people call something like `mbedtls_cipher_auth_xxx()` (the AEAD part of the Cipher API), it goes `cipher(GCM)->gcm->cipher(ECB)->aes/aria/camellia`. There's no `gcm->cipher(GCM)` arrow anywhere, and the `cipher(GMC)->gcm` part is not something we need in `CIPHER_LIGHT`.